// @rsbuild/core/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  createRsbuild: () => import_createRsbuild.createRsbuild,
  defineConfig: () => import_config.defineConfig,
  loadConfig: () => import_config.loadConfig,
  loadEnv: () => import_loadEnv.loadEnv,
  logger: () => import_shared.logger,
  mergeRsbuildConfig: () => import_shared.mergeRsbuildConfig,
  version: () => version
});
module.exports = __toCommonJS(src_exports);
var import_loadEnv = require("./loadEnv");
var import_createRsbuild = require("./createRsbuild");
var import_config = require("./cli/config");
var import_shared = require("@rsbuild/shared");
const version = "0.2.18";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createRsbuild,
  defineConfig,
  loadConfig,
  loadEnv,
  logger,
  mergeRsbuildConfig,
  version
});
