// @rspack/core/dist/exports.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CopyRspackPlugin = exports.SwcCssMinimizerRspackPlugin = exports.SwcJsMinimizerRspackPlugin = exports.HtmlRspackPlugin = exports.sharing = exports.container = exports.optimize = exports.webworker = exports.javascript = exports.wasm = exports.library = exports.electron = exports.node = exports.EnvironmentPlugin = exports.LoaderTargetPlugin = exports.LoaderOptionsPlugin = exports.HotModuleReplacementPlugin = exports.ExternalsPlugin = exports.EntryPlugin = exports.ProgressPlugin = exports.DefinePlugin = exports.ProvidePlugin = exports.BannerPlugin = exports.experimental_cleanupGlobalTrace = exports.experimental_registerGlobalTrace = exports.util = exports.config = exports.sources = exports.WebpackError = exports.Template = exports.ModuleFilenameHelpers = exports.NormalModule = exports.Stats = exports.RuntimeGlobals = exports.WebpackOptionsApply = exports.RspackOptionsApply = exports.MultiCompiler = exports.Compilation = exports.Compiler = exports.rspackVersion = exports.version = void 0;
const { version: rspackVersion, webpackVersion } = require("../package.json");
exports.rspackVersion = rspackVersion;
exports.version = webpackVersion;
var Compiler_1 = require("./Compiler");
Object.defineProperty(exports, "Compiler", { enumerable: true, get: function () { return Compiler_1.Compiler; } });
var Compilation_1 = require("./Compilation");
Object.defineProperty(exports, "Compilation", { enumerable: true, get: function () { return Compilation_1.Compilation; } });
var MultiCompiler_1 = require("./MultiCompiler");
Object.defineProperty(exports, "MultiCompiler", { enumerable: true, get: function () { return MultiCompiler_1.MultiCompiler; } });
const rspackOptionsApply_1 = require("./rspackOptionsApply");
Object.defineProperty(exports, "RspackOptionsApply", { enumerable: true, get: function () { return rspackOptionsApply_1.RspackOptionsApply; } });
Object.defineProperty(exports, "WebpackOptionsApply", { enumerable: true, get: function () { return rspackOptionsApply_1.RspackOptionsApply; } });
var RuntimeGlobals_1 = require("./RuntimeGlobals");
Object.defineProperty(exports, "RuntimeGlobals", { enumerable: true, get: function () { return RuntimeGlobals_1.RuntimeGlobals; } });
var Stats_1 = require("./Stats");
Object.defineProperty(exports, "Stats", { enumerable: true, get: function () { return Stats_1.Stats; } });
var NormalModule_1 = require("./NormalModule");
Object.defineProperty(exports, "NormalModule", { enumerable: true, get: function () { return NormalModule_1.NormalModule; } });
var ModuleFilenameHelpers_1 = require("./lib/ModuleFilenameHelpers");
Object.defineProperty(exports, "ModuleFilenameHelpers", { enumerable: true, get: function () { return __importDefault(ModuleFilenameHelpers_1).default; } });
var Template_1 = require("./Template");
Object.defineProperty(exports, "Template", { enumerable: true, get: function () { return __importDefault(Template_1).default; } });
exports.WebpackError = Error;
const sources = require("webpack-sources"); // use require to avoid wrong types, @types/webpack-sources is outdate
exports.sources = sources;
const config_1 = require("./config");
exports.config = {
    getNormalizedRspackOptions: config_1.getNormalizedRspackOptions,
    applyRspackOptionsDefaults: config_1.applyRspackOptionsDefaults,
    getNormalizedWebpackOptions: config_1.getNormalizedRspackOptions,
    applyWebpackOptionsDefaults: config_1.applyRspackOptionsDefaults
};
const createHash_1 = require("./util/createHash");
const cleverMerge_1 = require("./util/cleverMerge");
exports.util = { createHash: createHash_1.createHash, cleverMerge: cleverMerge_1.cachedCleverMerge };
var binding_1 = require("@rspack/binding");
Object.defineProperty(exports, "experimental_registerGlobalTrace", { enumerable: true, get: function () { return binding_1.registerGlobalTrace; } });
Object.defineProperty(exports, "experimental_cleanupGlobalTrace", { enumerable: true, get: function () { return binding_1.cleanupGlobalTrace; } });
///// Internal Plugins /////
var builtin_plugin_1 = require("./builtin-plugin");
Object.defineProperty(exports, "BannerPlugin", { enumerable: true, get: function () { return builtin_plugin_1.BannerPlugin; } });
var builtin_plugin_2 = require("./builtin-plugin");
Object.defineProperty(exports, "ProvidePlugin", { enumerable: true, get: function () { return builtin_plugin_2.ProvidePlugin; } });
var builtin_plugin_3 = require("./builtin-plugin");
Object.defineProperty(exports, "DefinePlugin", { enumerable: true, get: function () { return builtin_plugin_3.DefinePlugin; } });
var builtin_plugin_4 = require("./builtin-plugin");
Object.defineProperty(exports, "ProgressPlugin", { enumerable: true, get: function () { return builtin_plugin_4.ProgressPlugin; } });
var builtin_plugin_5 = require("./builtin-plugin");
Object.defineProperty(exports, "EntryPlugin", { enumerable: true, get: function () { return builtin_plugin_5.EntryPlugin; } });
var builtin_plugin_6 = require("./builtin-plugin");
Object.defineProperty(exports, "ExternalsPlugin", { enumerable: true, get: function () { return builtin_plugin_6.ExternalsPlugin; } });
var builtin_plugin_7 = require("./builtin-plugin");
Object.defineProperty(exports, "HotModuleReplacementPlugin", { enumerable: true, get: function () { return builtin_plugin_7.HotModuleReplacementPlugin; } });
var LoaderOptionsPlugin_1 = require("./lib/LoaderOptionsPlugin");
Object.defineProperty(exports, "LoaderOptionsPlugin", { enumerable: true, get: function () { return LoaderOptionsPlugin_1.LoaderOptionsPlugin; } });
var LoaderTargetPlugin_1 = require("./lib/LoaderTargetPlugin");
Object.defineProperty(exports, "LoaderTargetPlugin", { enumerable: true, get: function () { return LoaderTargetPlugin_1.LoaderTargetPlugin; } });
var EnvironmentPlugin_1 = require("./lib/EnvironmentPlugin");
Object.defineProperty(exports, "EnvironmentPlugin", { enumerable: true, get: function () { return EnvironmentPlugin_1.EnvironmentPlugin; } });
const NodeTemplatePlugin_1 = __importDefault(require("./node/NodeTemplatePlugin"));
const builtin_plugin_8 = require("./builtin-plugin");
exports.node = { NodeTargetPlugin: builtin_plugin_8.NodeTargetPlugin, NodeTemplatePlugin: NodeTemplatePlugin_1.default };
const builtin_plugin_9 = require("./builtin-plugin");
exports.electron = { ElectronTargetPlugin: builtin_plugin_9.ElectronTargetPlugin };
const builtin_plugin_10 = require("./builtin-plugin");
exports.library = { EnableLibraryPlugin: builtin_plugin_10.EnableLibraryPlugin };
const builtin_plugin_11 = require("./builtin-plugin");
exports.wasm = { EnableWasmLoadingPlugin: builtin_plugin_11.EnableWasmLoadingPlugin };
const builtin_plugin_12 = require("./builtin-plugin");
exports.javascript = { EnableChunkLoadingPlugin: builtin_plugin_12.EnableChunkLoadingPlugin };
const builtin_plugin_13 = require("./builtin-plugin");
exports.webworker = { WebWorkerTemplatePlugin: builtin_plugin_13.WebWorkerTemplatePlugin };
const builtin_plugin_14 = require("./builtin-plugin");
exports.optimize = { LimitChunkCountPlugin: builtin_plugin_14.LimitChunkCountPlugin };
const ModuleFederationPlugin_1 = require("./container/ModuleFederationPlugin");
const ModuleFederationPluginV1_1 = require("./container/ModuleFederationPluginV1");
const ContainerPlugin_1 = require("./container/ContainerPlugin");
const ContainerReferencePlugin_1 = require("./container/ContainerReferencePlugin");
exports.container = {
    ContainerPlugin: ContainerPlugin_1.ContainerPlugin,
    ContainerReferencePlugin: ContainerReferencePlugin_1.ContainerReferencePlugin,
    ModuleFederationPlugin: ModuleFederationPlugin_1.ModuleFederationPlugin,
    ModuleFederationPluginV1: ModuleFederationPluginV1_1.ModuleFederationPluginV1
};
const ProvideSharedPlugin_1 = require("./sharing/ProvideSharedPlugin");
const ConsumeSharedPlugin_1 = require("./sharing/ConsumeSharedPlugin");
const SharePlugin_1 = require("./sharing/SharePlugin");
exports.sharing = {
    ProvideSharedPlugin: ProvideSharedPlugin_1.ProvideSharedPlugin,
    ConsumeSharedPlugin: ConsumeSharedPlugin_1.ConsumeSharedPlugin,
    SharePlugin: SharePlugin_1.SharePlugin
};
///// Rspack Postfixed Internal Plugins /////
var builtin_plugin_15 = require("./builtin-plugin");
Object.defineProperty(exports, "HtmlRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_15.HtmlRspackPlugin; } });
var builtin_plugin_16 = require("./builtin-plugin");
Object.defineProperty(exports, "SwcJsMinimizerRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_16.SwcJsMinimizerRspackPlugin; } });
var builtin_plugin_17 = require("./builtin-plugin");
Object.defineProperty(exports, "SwcCssMinimizerRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_17.SwcCssMinimizerRspackPlugin; } });
var builtin_plugin_18 = require("./builtin-plugin");
Object.defineProperty(exports, "CopyRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_18.CopyRspackPlugin; } });
