// @rspack/core/dist/util/fake.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFakeCompilationDependencies = exports.createProcessAssetsHook = exports.createFakeHook = exports.createFakeProcessAssetsHook = void 0;
const tapable = __importStar(require("tapable"));
const MergeCaller_1 = __importDefault(require("./MergeCaller"));
const createFakeProcessAssetsHook = (compilation) => {
    const createFakeTap = (options, 
    // @ts-expect-error
    fn, tap) => {
        var _a;
        if (typeof options === "string")
            options = { name: options };
        const hook = compilation.__internal_getProcessAssetsHookByStage((_a = options.stage) !== null && _a !== void 0 ? _a : 0);
        // @ts-expect-error
        hook[tap](options.name, fn);
    };
    return {
        name: "processAssets",
        tap: (options, fn) => createFakeTap(options, fn, "tap"),
        tapAsync: (options, fn) => createFakeTap(options, fn, "tapAsync"),
        tapPromise: (options, fn) => createFakeTap(options, fn, "tapPromise"),
        stageAdditional: new tapable.AsyncSeriesHook(["assets"]),
        stagePreProcess: new tapable.AsyncSeriesHook(["assets"]),
        stageDerived: new tapable.AsyncSeriesHook(["assets"]),
        stageAdditions: new tapable.AsyncSeriesHook(["assets"]),
        stageNone: new tapable.AsyncSeriesHook(["assets"]),
        stageOptimize: new tapable.AsyncSeriesHook(["assets"]),
        stageOptimizeCount: new tapable.AsyncSeriesHook(["assets"]),
        stageOptimizeCompatibility: new tapable.AsyncSeriesHook(["assets"]),
        stageOptimizeSize: new tapable.AsyncSeriesHook(["assets"]),
        stageDevTooling: new tapable.AsyncSeriesHook(["assets"]),
        stageOptimizeInline: new tapable.AsyncSeriesHook(["assets"]),
        stageSummarize: new tapable.AsyncSeriesHook(["assets"]),
        stageOptimizeHash: new tapable.AsyncSeriesHook(["assets"]),
        stageOptimizeTransfer: new tapable.AsyncSeriesHook(["assets"]),
        stageAnalyse: new tapable.AsyncSeriesHook(["assets"]),
        stageReport: new tapable.AsyncSeriesHook(["assets"])
    };
};
exports.createFakeProcessAssetsHook = createFakeProcessAssetsHook;
const createFakeHook = (fakeHook) => {
    return Object.freeze(Object.assign(fakeHook, { _fakeHook: true }));
};
exports.createFakeHook = createFakeHook;
const createProcessAssetsHook = (processAssetsHooks, name, stage, getArgs) => {
    const errorMessage = (reason) => `Can't automatically convert plugin using Compilation.hooks.${name} to Compilation.hooks.processAssets because ${reason}.
BREAKING CHANGE: Asset processing hooks in Compilation has been merged into a single Compilation.hooks.processAssets hook.`;
    const getOptions = (options) => {
        if (typeof options === "string")
            options = { name: options };
        if (options.stage) {
            throw new Error(errorMessage("it's using the 'stage' option"));
        }
        return { ...options, stage: stage };
    };
    const tap = (options, fn) => {
        processAssetsHooks.tap(getOptions(options), () => fn(...getArgs()));
    };
    const tapAsync = (options, fn) => {
        processAssetsHooks.tapAsync(getOptions(options), (assets, callback) => fn(...getArgs(), callback));
    };
    const tapPromise = (options, fn) => {
        processAssetsHooks.tapPromise(getOptions(options), () => fn(...getArgs()));
    };
    return (0, exports.createFakeHook)({
        name,
        intercept() {
            throw new Error(errorMessage("it's using 'intercept'"));
        },
        tap,
        tapAsync,
        tapPromise
    });
};
exports.createProcessAssetsHook = createProcessAssetsHook;
function createFakeCompilationDependencies(getDeps, addDeps) {
    const addDepsCaller = new MergeCaller_1.default(addDeps, 10);
    return {
        *[Symbol.iterator]() {
            const deps = getDeps();
            for (const dep of deps) {
                yield dep;
            }
        },
        has(dep) {
            return getDeps().includes(dep);
        },
        add: (dep) => {
            addDepsCaller.push(dep);
        },
        addAll: (deps) => {
            addDepsCaller.push(...deps);
        }
    };
}
exports.createFakeCompilationDependencies = createFakeCompilationDependencies;
