// @rsbuild/core/dist/provider/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var provider_exports = {};
__export(provider_exports, {
  applyBaseCSSRule: () => import_css.applyBaseCSSRule,
  applyCSSModuleRule: () => import_css.applyCSSModuleRule,
  createContext: () => import_createContext.createContext,
  createPublicContext: () => import_createContext.createPublicContext,
  getHTMLPlugin: () => import_htmlPluginUtil.getHTMLPlugin,
  getPluginAPI: () => import_initPlugins.getPluginAPI,
  initHooks: () => import_initHooks.initHooks,
  initPlugins: () => import_shared.initPlugins,
  initRsbuildConfig: () => import_initConfigs.initRsbuildConfig,
  rspackProvider: () => import_provider.rspackProvider,
  setHTMLPlugin: () => import_htmlPluginUtil.setHTMLPlugin,
  withDefaultConfig: () => import_config.withDefaultConfig
});
module.exports = __toCommonJS(provider_exports);
var import_provider = require("./provider");
var import_createContext = require("./core/createContext");
var import_shared = require("@rsbuild/shared");
var import_initHooks = require("./core/initHooks");
var import_config = require("./config");
var import_initConfigs = require("./core/initConfigs");
var import_initPlugins = require("./core/initPlugins");
var import_css = require("./plugins/css");
var import_htmlPluginUtil = require("./htmlPluginUtil");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  applyBaseCSSRule,
  applyCSSModuleRule,
  createContext,
  createPublicContext,
  getHTMLPlugin,
  getPluginAPI,
  initHooks,
  initPlugins,
  initRsbuildConfig,
  rspackProvider,
  setHTMLPlugin,
  withDefaultConfig
});
