// @rsbuild/core/dist/provider/provider.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var provider_exports = {};
__export(provider_exports, {
  rspackProvider: () => rspackProvider
});
module.exports = __toCommonJS(provider_exports);
var import_shared = require("@rsbuild/shared");
var import_createContext = require("./core/createContext");
var import_initConfigs = require("./core/initConfigs");
var import_initPlugins = require("./core/initPlugins");
var import_shared2 = require("./shared");
const rspackProvider = async ({
  pluginStore,
  rsbuildOptions,
  plugins
}) => {
  const rsbuildConfig = (0, import_shared.pickRsbuildConfig)(rsbuildOptions.rsbuildConfig);
  const context = await (0, import_createContext.createContext)(rsbuildOptions, rsbuildConfig, "rspack");
  const pluginAPI = (0, import_initPlugins.getPluginAPI)({ context, pluginStore });
  context.pluginAPI = pluginAPI;
  return {
    bundler: "rspack",
    pluginAPI,
    publicContext: (0, import_createContext.createPublicContext)(context),
    async applyDefaultPlugins() {
      pluginStore.addPlugins(await (0, import_shared2.applyDefaultPlugins)(plugins));
    },
    async createCompiler() {
      const { createCompiler } = await Promise.resolve().then(() => __toESM(require("./core/createCompiler")));
      const { rspackConfigs } = await (0, import_initConfigs.initConfigs)({
        context,
        pluginStore,
        rsbuildOptions
      });
      return createCompiler({
        context,
        rspackConfigs
      });
    },
    async getServerAPIs(options) {
      const { getServerAPIs } = await Promise.resolve().then(() => __toESM(require("../server/devServer")));
      const { createDevMiddleware } = await Promise.resolve().then(() => __toESM(require("./core/createCompiler")));
      await (0, import_initConfigs.initRsbuildConfig)({ context, pluginStore });
      return getServerAPIs(
        { context, pluginStore, rsbuildOptions },
        createDevMiddleware,
        options
      );
    },
    async startDevServer(options) {
      const { startDevServer } = await Promise.resolve().then(() => __toESM(require("../server/devServer")));
      const { createDevMiddleware } = await Promise.resolve().then(() => __toESM(require("./core/createCompiler")));
      await (0, import_initConfigs.initRsbuildConfig)({ context, pluginStore });
      return startDevServer(
        { context, pluginStore, rsbuildOptions },
        createDevMiddleware,
        options
      );
    },
    async preview(options) {
      const { startProdServer } = await Promise.resolve().then(() => __toESM(require("../server/prodServer")));
      await (0, import_initConfigs.initRsbuildConfig)({ context, pluginStore });
      return startProdServer(context, context.config, options);
    },
    async build(options) {
      const { build: buildImpl, rspackBuild } = await Promise.resolve().then(() => __toESM(require("./core/build")));
      return buildImpl(
        { context, pluginStore, rsbuildOptions },
        options,
        rspackBuild
      );
    },
    async initConfigs() {
      const { rspackConfigs } = await (0, import_initConfigs.initConfigs)({
        context,
        pluginStore,
        rsbuildOptions
      });
      return rspackConfigs;
    },
    async inspectConfig(inspectOptions) {
      const { inspectConfig } = await Promise.resolve().then(() => __toESM(require("./core/inspectConfig")));
      return inspectConfig({
        context,
        pluginStore,
        rsbuildOptions,
        inspectOptions
      });
    }
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  rspackProvider
});
