// @rsbuild/core/dist/server/getDevMiddlewares.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getDevMiddlewares_exports = {};
__export(getDevMiddlewares_exports, {
  getMiddlewares: () => getMiddlewares
});
module.exports = __toCommonJS(getDevMiddlewares_exports);
var import_url = __toESM(require("url"));
var import_middlewares = require("./middlewares");
var import_path = require("path");
const applySetupMiddlewares = (dev, compileMiddlewareAPI) => {
  const setupMiddlewares = dev.setupMiddlewares || [];
  const serverOptions = {
    sockWrite: (type, data) => compileMiddlewareAPI?.sockWrite(type, data)
  };
  const before = [];
  const after = [];
  setupMiddlewares.forEach((handler) => {
    handler(
      {
        unshift: (...handlers) => before.unshift(...handlers),
        push: (...handlers) => after.push(...handlers)
      },
      serverOptions
    );
  });
  return { before, after };
};
const applyDefaultMiddlewares = async ({
  middlewares,
  dev,
  compileMiddlewareAPI,
  output,
  pwd
}) => {
  const upgradeEvents = [];
  if (dev.compress) {
    const { default: compression } = await Promise.resolve().then(() => __toESM(require("../../compiled/http-compression")));
    middlewares.push((req, res, next) => {
      compression({
        gzip: true,
        brotli: false
      })(req, res, next);
    });
  }
  middlewares.push((req, res, next) => {
    res.setHeader("Access-Control-Allow-Origin", "*");
    const path = req.url ? import_url.default.parse(req.url).pathname : "";
    if (path?.includes("hot-update")) {
      res.setHeader("Access-Control-Allow-Credentials", "false");
    }
    const confHeaders = dev.headers;
    if (confHeaders) {
      for (const [key, value] of Object.entries(confHeaders)) {
        res.setHeader(key, value);
      }
    }
    next();
  });
  if (dev.proxy) {
    const { createProxyMiddleware } = await Promise.resolve().then(() => __toESM(require("./proxy")));
    const { middlewares: proxyMiddlewares, upgrade } = createProxyMiddleware(
      dev.proxy
    );
    upgradeEvents.push(upgrade);
    proxyMiddlewares.forEach((middleware) => {
      middlewares.push(middleware);
    });
  }
  if (compileMiddlewareAPI) {
    middlewares.push(compileMiddlewareAPI.middleware);
    upgradeEvents.push(
      compileMiddlewareAPI.onUpgrade.bind(compileMiddlewareAPI)
    );
  }
  if (dev.publicDir && dev.publicDir.name) {
    const { default: sirv } = await Promise.resolve().then(() => __toESM(require("../../compiled/sirv")));
    const { name } = dev.publicDir;
    const publicDir = (0, import_path.isAbsolute)(name) ? name : (0, import_path.join)(pwd, name);
    const assetMiddleware = sirv(publicDir, {
      etag: true,
      dev: true
    });
    middlewares.push(assetMiddleware);
  }
  const { distPath } = output;
  compileMiddlewareAPI && middlewares.push(
    (0, import_middlewares.getHtmlFallbackMiddleware)({
      distPath: (0, import_path.isAbsolute)(distPath) ? distPath : (0, import_path.join)(pwd, distPath),
      callback: compileMiddlewareAPI.middleware,
      htmlFallback: dev.htmlFallback
    })
  );
  if (dev.historyApiFallback) {
    const { default: connectHistoryApiFallback } = await Promise.resolve().then(() => __toESM(require("../../compiled/connect-history-api-fallback")));
    const historyApiFallbackMiddleware = connectHistoryApiFallback(
      dev.historyApiFallback === true ? {} : dev.historyApiFallback
    );
    middlewares.push(historyApiFallbackMiddleware);
    compileMiddlewareAPI?.middleware && middlewares.push(compileMiddlewareAPI.middleware);
  }
  middlewares.push(import_middlewares.faviconFallbackMiddleware);
  return {
    onUpgrade: (...args) => {
      upgradeEvents.forEach((cb) => cb(...args));
    }
  };
};
const getMiddlewares = async (options) => {
  const middlewares = [];
  const { compileMiddlewareAPI } = options;
  const { before, after } = applySetupMiddlewares(
    options.dev,
    compileMiddlewareAPI
  );
  before.forEach((fn) => middlewares.push(fn));
  const { onUpgrade } = await applyDefaultMiddlewares({
    middlewares,
    dev: options.dev,
    compileMiddlewareAPI,
    output: options.output,
    pwd: options.pwd
  });
  after.forEach((fn) => middlewares.push(fn));
  return {
    close: async () => {
      compileMiddlewareAPI?.close();
    },
    onUpgrade,
    middlewares
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getMiddlewares
});
